### ISSUE TITLE
self_test_malloc problem in core_self_tests

### ISSUE THREAD

**Issue by HeJyunYan** [OP-TEE/optee_os#3519]
Excuse me. I have a question about the malloc of bget.
When the pool of bget is not enough to split to two pools.
The rest pool will combine with the allocated pool. So the allocated pool will increase the size and not the size of request.
And the `malloc_buffer_is_within_alloced(p1 + 25, 1000)` of xtest-1001 will be failed.

```c
void *bget(requested_size, poolset)
   bufsize requested_size;
   struct bpoolset *poolset;
 {
        ...
        if ((b->bh.bsize - size) > (SizeQ + (sizeof(struct bhead)))) {
                ...
        } else {
                ...
                /* The buffer isn't big enough to split.  Give  the  whole
                        shebang to the caller and remove it from the free list. */
                ...
        }
        ...
}
```

**Comment by jenswi-linaro** [comment]
I don't understand. Do you have a fix to the problem too? If so please raise a pull request. I think it will be easier to understand the problem then.

**Comment by HeJyunYan** [comment]
Sorry. For example, if the remaining size of the memory pool is 1040 bytes,  and the requested size is 1024 bytes. Since 16 bytes (1040 - 1024) <= (SizeQ + (sizeof(struct bhead))), p1 will get 1040 bytes ,  it cause the test case failed.
It checks the range from addr 25 to 1025 and expect to get an error (addr 1025 is  not within the allocated range).
In fact, p1 got 1040 bytes actually. It will pass the check. And it results the case failed.

[Tue Jan  7 15:52:14.706 2020] [   59.237808] optee: E/TC:0   p1=0x0  p2=0x0  p3=0x0  p4=0x0

[Tue Jan  7 15:52:14.706 2020] [   59.243477] optee: E/TC:0 bget:702 DDDDD bget: malloc size is 418

[Tue Jan  7 15:52:14.706 2020] [   59.249714] optee: E/TC:0 bget:712 DDDDD bget: malloc addr is 0x9e444178, size is 408

[Tue Jan  7 15:52:14.706 2020] [   59.258336] optee: E/TC:0 - p1 = malloc(1024)

[Tue Jan  7 15:52:14.706 2020] [   59.262890] optee: E/TC:0 - p2 = malloc(1024)

[Tue Jan  7 15:52:14.706 2020] [   59.268162] optee: E/TC:0   p1=0x9e444178  p2=0x9e443368  p3=0x0  p4=0x0

[Tue Jan  7 15:52:15.077 2020] [   59.379251] optee: E/TC:0 dump_pool:980 [dump_pool][980] start_b:0x9e444178, size:0x410,

**Comment by HeJyunYan** [comment]
How to fix this test case? Any good idea :)

**Comment by jenswi-linaro** [comment]
This seems to be a corner case that can only occur when you're almost out of memory already. Or am I missing something?

**Comment by HeJyunYan** [comment]
> This seems to be a corner case that can only occur when you're almost out of memory already. Or am I missing something?

In fact, it may not only occur when almost out of memory. When the BestFit is enabled. It occurs when the remaining size (already minus the requested size) of the found buffer in poolset is less than (SizeQ + (sizeof(struct bhead))). And there are still enough memory in the poolset.

**Comment by HeJyunYan** [comment]
Sorry, I press the wrong key :(

**Comment by lenormandfranck** [comment]
The issue was hapenning due to bget behavior explained here: https://github.com/OP-TEE/optee_os/blob/27e8d08d664b80a632eac1065d4904f85220bd75/lib/libutils/isoc/bget.c#L644
Implemented by this code: https://github.com/OP-TEE/optee_os/blob/27e8d08d664b80a632eac1065d4904f85220bd75/lib/libutils/isoc/bget.c#L653

This code has been change by https://github.com/OP-TEE/optee_os/commit/cc5981b239d3f1d1ef4cffa3d8cdd407c4c7ef79#diff-41358c5e0222a6f2bcdfb41a5a93a2bc09a4ae5bc97e36ec6c37695115a9c079 which splits the buffer to remove the part not requested. 

TLDR: Bget was allocating a whole free block if the space left in the block was not enough for a future allocation. This is no longer the case.
